### PR TITLE
Initial pipes implementation

### PR DESCRIPTION
**Pull request by mrkno**
Resolves https://github.com/concierge/Concierge/issues/249

### CODE REVIEW

File: core/modules/loopback.js [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+    }
+
+    sendMessage(message, thread) {
+        this._loopback(message, thread);
+        if (this._pipeEvents.length === 0) {
+            return this._baseApi.sendMessage(message, thread);
+        }
+    }
+
+    sendUrl(url, thread) {
```
Comment (Awarua-) [inline core/modules/loopback.js]: I see that it only will respond when using sendUrl or sendMessage are there other api methods that should also be looped back on, such as sendPrivateMessage, setTitle?

File: core/modules/loopback.js [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+        const pipeEvents = [];
+        if (currConfig.pipe !== false && event.body.contains('|')) {
+            const newBodies = event.body.split(/\s*\|\s*/g);
+            for (let n of newBodies) {
+                const e = shim.createEvent(event.thread_id, event.sender_id, event.sender_name, n);
+                e.event_source = event.event_source;
+                pipeEvents.push(e);
+            }
+        }
+        const loopDepth = currConfig.enabled ? currConfig.maxDepth || platform.loopDepth || 0 : 0;
```
Comment (Awarua-) [inline core/modules/loopback.js]: confusing to read, but I understand

Comment (mrkno) [inline core/modules/loopback.js]: One liner, easy enough to change if you would like

File: core/modules/loopback.js (hunk 1 above)
Comment (mrkno) [inline core/modules/loopback.js]: - setTitle doesn't make sense to intercept, as its not a message as such and precisely what is sent is very implementation specific.
- sendPrivateMessage is just that, private. Additionally it does not have a thread so the output will be piped to someones private chat.
- sendFile and sendImage either fallback to sendUrl or provide implementations for sending a byte stream which isn't really something we are setup to process receiving
- the other methods are really irrelevant

**Comment by mrkno** [comment]
@Awarua- I agree it is hard logic to follow. Its a non-optimal solution to the problem where the best way to deal with it would be to get the integrations themselves to implement loopback and piping. That isn't going to happen because then everything would be inconsistent.

Instead this works by implementing the decorator pattern, where the decorator generates new decorators on the fly as needed (so we can keep track of recursion depth while still allowing commands at the current recursion depth). If you have any better ways I'm open to suggestions?

File: core/modules/loopback.js (hunk 1 above)
Comment (Awarua-) [inline core/modules/loopback.js]: Ok fair enough

**Comment by Awarua-** [comment]
I don't have a better solution, at this time.

**Pull request merged**